frontend/src/metabase/visualizations/components/Visualization.jsx

```diff
@@ -84,35 +88,60 @@ export default class Visualization extends Component {
     }
 
     componentWillReceiveProps(newProps) {
-        if (isSameSeries(newProps.series, this.props.series)) {
-            // clear the error so we can try to render again
-            this.setState({ error: null });
-        } else {
+        if (!isSameSeries(newProps.series, this.props.series) || !Utils.equals(newProps.settings, this.props.settings)) {
             this.transform(newProps);
         }
     }
 
+    componentDidMount() {
+        this.updateWarnings();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (!Utils.equals(this.getWarnings(prevProps, prevState), this.getWarnings())) {
+            this.updateWarnings();
+        }
+    }
+
+    getWarnings(props = this.props, state = this.state) {
```

Comment:
I'm assuming you did this to provide a nicer shorthand for props and state inside the function? I'm a fan. I did notice you didn't use the same pattern in some of the other non lifecycle component methods so I'm curious if there was a particular reason it was useful here.